src/api/actions/uploadInBackground.ts

```diff
@@ -0,0 +1,134 @@
+
+export const uploadInBackground = async (
+  apiUri: string,
+  urls: string[],
+  uploadedCallback?: (url: string, ipfsUrl: string) => void
+): Promise<UrlToIPFS> => {
+  let completedJobs: UrlToIPFS = {};
+  const urlBatches: string[][] = chunkArray(urls, URLS_PER_CHUNK_START);
+  const startBulkResponses: UrlToJobId = (
+    await Promise.all(
+      // .map, .foreach cannot be async
+      // to get around, you have to make it return a promise
+      urlBatches.map(async (urls): Promise<UrlToJobId> => {
+        return await tryStartBulkUpload(apiUri, urls);
+      })
+    )
+  ).reduce((accumulator, currentResponse) => {
+    Object.entries(currentResponse).forEach(
+      ([key, value]) => (accumulator[key] = value)
+    );
+    return accumulator;
+  });
+  for (const key of Object.keys(startBulkResponses)) {
+    incompleteJobs[startBulkResponses[key]] = key;
+  }
+  // same idea as above, map into promise and then await all promises
+  while (Object.keys(incompleteJobs).length != 0) {
+    let jobIdChunks: string[][] = chunkArray(
+      Object.keys(incompleteJobs),
+      URLS_PER_CHUNK_CHECK
+    );
+
+    for (const jobIds of jobIdChunks) {
+      const statuses = await tryCheckBulkUploadJob(apiUri, jobIds);
+      for (const jobId of jobIds) {
+        const status = statuses[jobId];
+        if (status?.isCompleted ?? false) {
+          if (status.failed) {
+            failedCheckJobs[jobId] = status;
+          } else {
+            completedJobs[jobId] = status.result.hash;
+            uploadedCallback?.(jobId, status.result.hash);
+          }
+          delete incompleteJobs[jobId];
+        }
+      }
+    }
+    await delay(10);
+  }
+
+  fs.writeFileSync(ERROR_FILEPATH, JSON.stringify(errorJobs ?? ""));
+  fs.writeFileSync(FAILURE_FILEPATH, JSON.stringify(failedCheckJobs ?? ""));
+
+  return completedJobs;
+};
+
+const delay = (milliseconds: number) => {
+  return new Promise((resolve) => setTimeout(resolve, milliseconds));
+};
+
+export const tryStartBulkUpload = async (
+  apiUri: string,
+  urls: string[]
+): Promise<UrlToJobId> => {
+  let uploadResponse: UrlToJobId = {};
+  try {
+    uploadResponse = await startBulkUpload(apiUri, urls);
+  } catch (e) {
+    urls.forEach((url) => {
+      delete incompleteJobs[uploadResponse[url]];
+      errorJobs.start.push({ url: url, error: e });
+    });
+  }
+  return uploadResponse;
+};
+
+export const tryCheckBulkUploadJob = async (
+  apiUri: string,
+  jobIds: string[]
+): Promise<UploadJobStatus> => {
+  let checkResponse: UploadJobStatus = {};
+  try {
+    checkResponse = await checkBulkUploadJob(apiUri, jobIds);
+  } catch (e) {
+    jobIds.forEach((jobId) => {
+      delete incompleteJobs[jobId];
+      errorJobs.check.push({
+        url: incompleteJobs[jobId] ?? "unknown",
+        jobId,
+        error: e,
+      });
+    });
+  }
+
+  return checkResponse;
+};
+
+const chunkArray = (array: any[], chunkSize: number): any[][] => {
```

Comment:
I hear you, but this variable described the thing being chunked, so I don't think `chunks` is an improvement over `array`. Open to other names though!

Reply:
`array` to me is generic because it has no indication of what is in the array, it's just a list of unknown things. `chunks` or something similar is an improvement because it indicates what is in the array.

Like I said just a nitpick though, and the variable itself is `chunkArray` so it's not a blocking issue

Reply:
Perhaps `original` or `input`